### DISCUSSION TITLE
How to set custom quality when resizing on command?

### DISCUSSION THREAD

**Discussion by ajtatum** [SixLabors/ImageSharp.Web#123] [Q&A] [answered]
Hi,

I use ImageSharp to generally optimize any image a user uploads, for example, their profile photo. 

In the first photo, the photo was resized to 512x512 and setting the Jpeg quality to 80%. The second picture was dynamically resized to be 100px high, and, as you can see... the quality isn't the greatest. I noticed this with ImageSharp in general when making thumbnails, which is why I usually set the quality higher for thumbnails and, in some cases, apply GaussianSharpenProcessor(2). So, is there a way I can set the quality of the image served dynamically?

Thank you,
AJ

![ce2dbc4a-7376-45d6-bbbc-5d80d3f56014](https://user-images.githubusercontent.com/318022/93284580-fe35fb00-f7a0-11ea-93ab-3b8b48abe193.jpg)

![ce2dbc4a-7376-45d6-bbbc-5d80d3f56014-100h](https://user-images.githubusercontent.com/318022/93284582-fece9180-f7a0-11ea-9457-3cf1ed1db30a.jpg)

**Comment by JimBobSquarePants** [comment]
ImageSharp or ImageSharp.Web?

**Reply by ajtatum** [reply]
ImageSharp.Web (whichever is the middle where). The thumbnail was generated by adding a query string "?height=100"

**Reply by JimBobSquarePants** [reply]
ImageSharp.Web is the middleware (_The name of this repository also_)

> I noticed this with ImageSharp in general when making thumbnails, which is why I usually set the quality higher for thumbnails and, in some cases, apply GaussianSharpenProcessor(2)

This was why I was asking since `GuassianSharpen` (_Use the extension methods, don't call processors directly_) is not available out of the box for the middleware. 

You can set the default quality and subsampling for the jpg encoder by assigning a new `JpegEncoder` instance to the `JpegFormat` 

``` c#
configuration.ImageFormatsManager.SetEncoder(JpegFormat.Instance, new JpegEncoder());
```

If you want to do something like this on the fly you would have to implement and register your own processor extension for ImageSharp.Web

**Reply by tocsoft** [reply]
this makes my think there might be a place for a nice hook in here for allowing people to return custom encoders per request then they ca use what ever logic they desire based on parsed commands or query string args or anything else they choose to return different configurations each time... could even return totally different formats and we can then  use that encoder/settings for saving the cached image.

**Reply by JimBobSquarePants** [reply]
We can already request a different format but that loads the format instance from the registered configuration. Quality is something that's very specific to individual image formats so I deliberately chose not to implement anything out of the box. 

I don't want users to be able to pass per-request configurations but it should be possible to allow setting the encoder per request to match the ability to set the format per request. That would then make it pretty easy to set the jpeg quality per-request out of the box.

**Reply by tocsoft** [reply]
yeah I agree encoders was what I was thinking.

**Reply by JimBobSquarePants** [reply]
@ajtatum @tocsoft See #125

**Reply by ajtatum** [reply]
Awesome guys! Thanks!

**Comment by JimBobSquarePants** [comment, marked as answer]
This functionality now exists in V1. Thanks everyone!
